src/ethereum_optimized/byzantium/state_db.py

```diff
@@ -117,749 +75,210 @@ def __eq__(self, other: object) -> bool:
-    """
-    if state._tx_stack[-1] is not None:
-        raise Exception("Current transaction is a db transaction")
-    for (address, account) in state._dirty_accounts.pop().items():
-        state._dirty_accounts[-1][address] = account
-    state._destroyed_accounts[-2] |= state._destroyed_accounts[-1]
-    for address in state._destroyed_accounts.pop():
-        state._dirty_storage[-2].pop(address, None)
-    for (address, cache) in state._dirty_storage.pop().items():
-        if address not in state._dirty_storage[-1]:
-            state._dirty_storage[-1][address] = {}
-        for (key, value) in cache.items():
-            state._dirty_storage[-1][address][key] = value
-    state._tx_stack.pop()
-
-
-def rollback_transaction(state: State) -> None:
+def storage_root(state: State, address: Address) -> Root:
     """
     See `ethereum.byzantium.state`.
     """
-    if state._tx_stack[-1] is not None:
-        raise Exception("Current transaction is a db transaction")
-    state._tx_stack.pop()
-    state._dirty_accounts.pop()
-    state._dirty_storage.pop()
-    state._destroyed_accounts.pop()
-
-
-def get_internal_key(key: Bytes) -> Bytes:
-    """
-    Convert a key to the form used internally inside the trie.
-    """
-    return bytes_to_nibble_list(keccak256(key))
+    if len(state.tx_restore_points) != 0:
+        raise Exception("In a non-db transaction")
+    flush(state)
+    return state.db.storage_root(address)
 
 
-def get_storage(state: State, address: Address, key: Bytes) -> U256:
+def flush(state: State) -> None:
     """
-    See `ethereum.byzantium.state`.
+    Send everything in the internal caches to the Rust layer.
     """
-    for i in range(len(state._tx_stack) - 1, -1, -1):
-        if address in state._dirty_storage[i]:
-            if key in state._dirty_storage[i][address]:
-                cached_res = state._dirty_storage[i][address][key]
-                if cached_res is None:
-                    return U256(0)
-                else:
-                    return cached_res
-            if key in state._destroyed_accounts[i]:
-                # Higher levels refer to the account prior to destruction
-                return U256(0)
-    res = state._current_tx.get(b"\x01" + address + b"\x00" + key)
-    if res is None:
-        return U256(0)
-    else:
-        res = rlp.decode(res)
-        assert isinstance(res, bytes)
-        return U256.from_be_bytes(res)
+    if len(state.tx_restore_points) != 0:
+        raise Exception("In a non-db transaction")
+    for address in state.destroyed_accounts:
+        state.db.destroy_storage(address)
+    for address, account in state.dirty_accounts.items():
+        state.db.set_account(address, account)
+    for address, storage in state.dirty_storage.items():
+        for key, value in storage.items():
+            state.db.set_storage(address, key, value)
+    state.destroyed_accounts = set()
+    state.dirty_accounts = {}
+    state.dirty_storage = {}
```

Comment:
Would using `.clear()` be better for these? I think it causes less (de)allocations.